### ISSUE TITLE
Missing dependencies in esp-html-skeleton (6.0.0) (EDIT: Dependency needs to be manually installed)

### ISSUE THREAD

**Issue by lss4** [embedthis/esp#101]
It seems with `esp-html-skeleton` 6.0.0 the issue (#81, #84) reappeared.

Running Manjaro with ESP built (`make` with no flags) and installed (along with MakeMe, Pak, Expansive). I'm also using Appweb.

ESP: 8.1.0
Appweb: 7.2.1
Pak: 1.0.4
MakeMe: 1.0.2
Expansive: 1.0.1

Following the tutorial, it at first complained about missing `exp-js` dependency (as well as other `exp-xxx` dependencies). These can be manually installed one by one with `pak`. However, during the process I discovered some of these `exp-xxx` dependencies (for now it's `exp-js`, `exp-less`, `exp-css`) have `npm` dependencies which needs to be resolved as well. 

The `npm` is not mentioned in the Application Tour but it's mentioned in SPA Tour (for VueJS). Note the `npm` dependencies mentioned here might not be all of them as some might have been installed beforehand.

`exp-js`: `uglify-js`
`exp-less`: `less`
`exp-css`: `autoprefixer`

These `npm` dependencies need to be installed globally for once (via `npm install -g`, as the pak install would call, which requires root).

When instaling `exp-canon`, I saw both 1.0.0 and 0.3.0 appearing in the output.

```
     [Cache] exp-canon 1.0.0
     [Mkdir] /home/user/.paks/exp-canon/embedthis/1.0.0
       [Get] https://github.com/embedthis/exp-canon/archive/v1.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/exp-canon/embedthis/1.0.0
      [Info] exp-canon 0.3.0 successfully cached
   [Install] exp-canon 0.3.0
```

EDIT: After manually installing those dependencies beforehand (using `pak install embedthis/exp-js embedthis/exp-css embedthis/exp-less embedthis/exp-html embedthis/exp-esp embedthis/exp-canon`), I could successfully pull the skeleton. However, according to the docs, it should be able to pull the dependencies on its own.

And I'm still not certain about the version of exp-canon, as it looked for v1.0.0 but what I get is 0.3.0.

EDIT 2: Trying `expansive render` I got this.

```
      [Info] Using profile: dev
    [Render] contents/css/all.css.less
     [Error] Cannot render css/all.css for document dist/css/all.css
Cannot render: css/all.css.less
Error Exception: Cannot find postcss
Stack:
 [00] paks/exp-css/expansive.es, line 96, --fun_486-519186533--, throw new Error('Cannot find postcss')
 [01] src/expansive.es, line 1450, transform, contents = transform.render.call(this, contents, meta, transform)
 [02] src/expansive.es, line 1411, pipeline, contents = transform(contents, mapping, path, meta)
 [03] src/expansive.es, line 1364, renderContents, contents = pipeline(contents, meta)
 [04] src/expansive.es, line 1340, renderDocument, contents = renderContents(contents, meta)
 [05] src/expansive.es, line 1110, renderDocuments, renderDocument(file, meta)
 [06] src/expansive.es, line 964, render, renderDocuments()
 [07] src/expansive.es, line 636, process, render()
 [08] src/expansive.es, line 2107, __initializer__, expansive.process()


    [Render] contents/css/app.less
    [Render] contents/css/theme.less
      [Post] esp
      [Info] Rendered 0 files to "dist". Elapsed time 0.35 secs.
expansive: ERROR: Error Render had 1 errors
```

And it exits here. Installing postcss locally (`$ npm install`) or installing it globally (`# npm install -g`) does not help, either...

**Comment by dev-embedthis** [comment]
The version confusion was because some paks had dependency versions that were old.

We've republished some of the dependent paks with 1.0.0 versions. We've also added code to exp-js, exp-css, exp-less, exp-marked, etc to install the NPM required tools like uglify-js, node-sass, etc.

If you have node installed, it should now work more smoothly. If not, please look at the ~/.paks/NAME/embedthis/*.es script. You can probably hack it to work and post a patch here. 

Thanks.

**Comment by lss4** [comment]
I noticed some paks have been republished with actual 1.0.0 version, while some others still have an older actual version.

```
$ pak install embedthis/esp-html-skeleton
     [Cache] esp-html-skeleton 6.0.0
     [Mkdir] /home/user/.paks/esp-html-skeleton/embedthis/6.0.0
       [Get] https://github.com/embedthis/esp-html-skeleton/archive/v6.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/esp-html-skeleton/embedthis/6.0.0
     [Cache] esp-mvc 6.0.0
     [Mkdir] /home/user/.paks/esp-mvc/embedthis/6.0.0
       [Get] https://github.com/embedthis/esp-mvc/archive/v6.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/esp-mvc/embedthis/6.0.0
     [Cache] exp-js 1.0.0
     [Mkdir] /home/user/.paks/exp-js/embedthis/1.0.0
       [Get] https://github.com/embedthis/exp-js/archive/v1.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/exp-js/embedthis/1.0.0
      [Info] exp-js 1.0.0 successfully cached
     [Cache] exp-less 1.0.0
     [Mkdir] /home/user/.paks/exp-less/embedthis/1.0.0
       [Get] https://github.com/embedthis/exp-less/archive/v1.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/exp-less/embedthis/1.0.0
      [Info] exp-less 1.0.0 successfully cached
     [Cache] exp-css 1.0.0
     [Mkdir] /home/user/.paks/exp-css/embedthis/1.0.0
       [Get] https://github.com/embedthis/exp-css/archive/v1.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/exp-css/embedthis/1.0.0
      [Info] exp-css 1.0.0 successfully cached
     [Cache] exp-esp 1.0.0
     [Mkdir] /home/user/.paks/exp-esp/embedthis/1.0.0
       [Get] https://github.com/embedthis/exp-esp/archive/v1.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/exp-esp/embedthis/1.0.0
      [Info] exp-esp 0.3.1 successfully cached
     [Cache] exp-html 1.0.0
     [Mkdir] /home/user/.paks/exp-html/embedthis/1.0.0
       [Get] https://github.com/embedthis/exp-html/archive/v1.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/exp-html/embedthis/1.0.0
      [Info] exp-html 0.2.0 successfully cached
     [Cache] exp-canon 1.0.0
     [Mkdir] /home/user/.paks/exp-canon/embedthis/1.0.0
       [Get] https://github.com/embedthis/exp-canon/archive/v1.0.0.tar.gz
   [Extract] Extract to /home/user/.paks/exp-canon/embedthis/1.0.0
      [Info] exp-canon 0.3.0 successfully cached
      [Info] esp-mvc 6.0.0 successfully cached
      [Info] esp-html-skeleton 6.0.0 successfully cached
   [Install] exp-js 1.0.0
   [Install] exp-less 1.0.0
   [Install] exp-css 1.0.0
   [Install] exp-esp 0.3.1
   [Install] exp-html 0.2.0
   [Install] exp-canon 0.3.0
   [Install] esp-mvc 6.0.0
   [Install] esp-html-skeleton 6.0.0
```

As for the permission issue, following [this instructions](https://wiki.archlinux.org/index.php/Node.js_#Allow_user-wide_installations) I could at least get `npm install -g` usable without `sudo` to install packages user-wide (instead of system-wide which requires root).

However I'm still stuck at the point where it complains about missing `postcss`, and installing `postcss` via `npm` user-wide this time does not fix it either.

```
$ expansive render
      [Info] Using profile: dev
    [Render] contents/css/all.css.less
     [Error] Cannot render css/all.css for document dist/css/all.css
Cannot render: css/all.css.less
Error Exception: Cannot find postcss
Stack:
 [00] paks/exp-css/expansive.es, line 98, --fun_486-602576783--, throw new Error('Cannot find postcss')
 [01] src/expansive.es, line 1450, transform, contents = transform.render.call(this, contents, meta, transform)
 [02] src/expansive.es, line 1411, pipeline, contents = transform(contents, mapping, path, meta)
 [03] src/expansive.es, line 1364, renderContents, contents = pipeline(contents, meta)
 [04] src/expansive.es, line 1340, renderDocument, contents = renderContents(contents, meta)
 [05] src/expansive.es, line 1110, renderDocuments, renderDocument(file, meta)
 [06] src/expansive.es, line 964, render, renderDocuments()
 [07] src/expansive.es, line 636, process, render()
 [08] src/expansive.es, line 2107, __initializer__, expansive.process()


    [Render] contents/css/app.less
    [Render] contents/css/theme.less
    [Render] contents/index.esp
    [Render] index.esp
      [Post] esp
       [Run] esp compile dist/index.esp
      [Info] Rendered 1 files to "dist". Elapsed time 16.33 secs.
expansive: ERROR: Error Render had 1 errors
```

**Comment by dev-embedthis** [comment]
Can you try to nail this down.  Remove all esp-* and exp-* from your ~/.paks directory and retry. Then try to find which pak is referencing the old versions. This does not happen for us here.

**Comment by lss4** [comment]
I already did that (`rm -rf`'d everything in `~/.paks`). Otherwise it wouldn't be redownloaded.

I'll try this again.

**Comment by dev-embedthis** [comment]
So to diagnose, you need to find the offending version in your paks/*/pak.json files. Which pak is referencing an old version.

**Comment by lss4** [comment]
It's not that it's referencing an old version. The archive `pak` downloaded is correct (`v1.0.0.tar.gz`). It's that the `version` field in `pak.json` is wrong (earlier).

e.g. `pak.json` of `exp-canon`:

```
{
    "name": "exp-canon",
    "description": "Canonicalize index pages for search engines",
    "version": "0.3.0", --> Notice here
    "keywords": [
        "expansive-plugin",
        "canon"
    ],
    "licenses": "GPL",
    "repository": {
        "type": "git",
        "url": "git://github.com/embedthis/exp-canon.git"
    },
    "devDependencies": {
        "expansive": ">=0.7.0",
        "pak": ">=0.11.3"
    },
    "origin": "embedthis/exp-canon"
}
```

**Comment by dev-embedthis** [comment]
My pak looks diferent:

``
{
    "name": "exp-canon",
    "description": "Canonicalize index pages for search engines",
    "version": "1.0.0",
    "keywords": [
        "expansive-plugin",
        "canon"
    ],
    "licenses": "GPL",
    "repository": {
        "type": "git",
        "url": "git://github.com/embedthis/exp-canon.git"
    },
    "devDependencies": {
        "expansive": ">=1.0.0",
        "pak": ">=1.0.0"
    },
    "origin": "embedthis/exp-canon"
}
``

**Comment by lss4** [comment]
Can you confirm whether the ones hosted on GitHub releases is really the correct one?

For example, I just checked exp-canon's [GitHub releases](https://github.com/embedthis/exp-canon/releases), and it seems both 1.0.0 and 0.3.0 are pointing to [the same commit](https://github.com/embedthis/exp-canon/commit/bbbd6418ba4baa7ed7ee33787bdb9ec93a14b905).

**Comment by dev-embedthis** [comment]
Thanks -- that's it.

Proper version just committed.

**Comment by lss4** [comment]
Okay, it seems packages have been updated to the correct version number.

As for the issues such as `Cannot find postcss`... it seems `exp-css` needs not only `postcss` and `autoprefixer`, but also their CLI counterparts (`postcss-cli` and `autoprefixer-cli`).

**Comment by dev-embedthis** [comment]
Fix checked in.

**Comment by lss4** [comment]
Okay.

It might be better if `pak.json` can be adjusted to pull everything needed as well.

Additionally, it might be better if the `npm install -g` command in expansive.es can be run unconditionally before actually checking the presence of the command, so if the command is still not found (or other errors occur), it would mean an error on the npm side that the user needs to check.

Here's an example patch (based on 782d702703fa5272291d1847006686c6af09ac67). It covers the changes that I did for pak.json and expansive.es.
[exp-css-dependency-check.patch.txt](https://github.com/embedthis/esp/files/4176450/exp-css-dependency-check.patch.txt)

As for the expansive.es changes. I just tested `expansive render` with everything (postcss/postcss-cli, autoprefixer/autoprefixer-cli and less/lessc) uninstalled (so they're initially "not present" in the system) and the process could still succeed without error (as the dependencies have been successfully installed via `npm` during the render process).
